dojo_plugin/pages/canvas.py

```diff
@@ -0,0 +1,225 @@
+
+""" 
+This function will retrieve the grade information, format it, and send to canvas
+However, if both user_id and module_id are supplied it will do a single user/module update 
+"""
+def do_canvas_sync(dojo, user_id=None, module_id=None):
+    
+    canvas_token = dojo.course.get("canvas_token","")
+    
+    canvas_api_host = dojo.course.get("canvas_api_host","")
+    canvas_course_id = dojo.course.get("canvas_course_id",0)
+    if len(canvas_token) == 0:
+        return {"status": "completed", "message": "Missing canvas_token in course.xml"}
+    if len(canvas_api_host) == 0:
+        return {"status": "completed", "message": "Missing canvas_api_host in course.xml"}
+    if canvas_course_id == 0:
+        return {"status": "completed", "message": "Missing canvas_course_id in course.xml"}
+    
+    canvas_grade_data = {}
+    
+    ignore_pending = request.args.get("ignore_pending") is not None
+        
+    students = {student.user_id: student.token for student in dojo.students}
+    course_students = dojo.course.get("students", [])
+    missing_students = list(set(course_students) - set(students.values()))
+    
+    # if user_id is None, it's a bulk grade submission, get all users, else just get user_id
+    if user_id is None: 
+        users = (
+                Users
+                .query
+                .join(DojoStudents, DojoStudents.user_id == Users.id)
+                .filter(DojoStudents.dojo == dojo,
+                        DojoStudents.token.in_(course_students))
+            )
+    else:
+        # get the user_id
+        users = (
+                Users
+                .query
+                .join(DojoStudents, DojoStudents.user_id == Users.id)
+                .filter(DojoStudents.dojo == dojo,
+                        DojoStudents.token.in_(course_students),
+                        DojoStudents.user_id == user_id)
+            )
+        if users.count() == 0:
+            return {"status": "completed", "message": f"Student {user_id} has not linked their student id with this course"}
+
+    grades = sorted(grade(dojo, users, ignore_pending=ignore_pending),
+                        key=lambda grade: grade["overall_grade"],
+                        reverse=True)
+    assessments = dojo.course.get("assessments", [])
+        
+    posting_results = []
+    progress_urls = []
+    assessment_student_counter = {}
+    for aindex, assessment in enumerate(assessments):
+        if "canvas_assignment_id" not in assessment:
+            continue
+        assessment_student_counter[assessment['id']] = 0
+        # if we have a value module_id, check for it and skip all the others
+        if module_id is not None and module_id != assessment["id"]:                
+            continue 
+                    
+        canvas_assignment_id = assessment["canvas_assignment_id"]
+        for grade_res in grades:                
+            grade_credit_percent = f"{grade_res['assessment_grades'][aindex]['credit'] * 100:.2f}%"
```

Comment:
How does the `.2f` map with what the student sees on the grade page in pwn.college? I'm just curious / worried about rounding issues and any discrepancies between what's on pwn.college and what's in canvas will cause problems.

Reply:
The display for admins and students use the templates grades_admin.html and course.html they both use %2f